lib/apipie_bindings/action.rb

```diff
@@ -57,35 +55,40 @@ def find_route(params={})
       return suitable_route
     end
 
-    def validate!(params)
-      missing_params = missing_params(api_params_tree { |par| par.required? }, params_tree(params))
-      raise ApipieBindings::MissingArgumentsError.new(missing_params) unless missing_params.empty?
-    end
-
-    def api_params_tree(&block)
-      ApipieBindings::Utilities.params_hash_tree(self.params, &block)
-    end
-
-    def params_tree(params)
-      params.inject([]) do |tree, (key, val)|
-        subtree = val.is_a?(Hash) ? { key.to_s => params_tree(val) } : key.to_s
-        tree << subtree
-        tree
-      end
+    def validate!(parameters)
+      errors = validate(params, parameters)
+      raise ApipieBindings::MissingArgumentsError.new(errors) unless errors.empty?
     end
 
-    def missing_params(master, slave)
-      missing = []
-      master.each do |required_param|
-        if required_param.is_a?(Hash)
-          key = required_param.keys.first
-          slave_hash = slave.select { |p| p.is_a?(Hash) && p[key] }
-          missing << missing_params(required_param[key], slave_hash.first ? slave_hash.first[key] : [])
-        else
-          missing << required_param unless slave.include?(required_param)
+    def validate(level_params, level_given, path='%s')
+      # check required
+      required_params = level_params.select(&:required?).map(&:name)
+      given_params = level_given.keys.select { |par| !level_given[par].nil? }.map(&:to_s)
+      missing_params = required_params - given_params
+      missing_params.map! { |p| path % p }
+
+      # check individuals one by one
+      level_given.each do |param, value|
+        param_description = level_params.find { |p| p.name == param.to_s }
+        if param_description
+        
+          # nested?
+          if !param_description.params.empty? && !value.nil?
+            # array
+            if param_description.expected_type == :array
+              value.each.with_index do |item, i|
+                missing_params += validate(param_description.params, item, path=(path % "#{param_description.name}" + "[#{i}][%s]"))
```

Comment:
But wouldn't the recursive validate call fail on this line https://github.com/Apipie/apipie-bindings/pull/26/files#diff-e9a2c135f011b8a56a1cd6f71f92d298R66 because the item would be `integer`, that hasn't the `keys` method